src/pmdas/lustre/pmdalustre.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2011 SGI.  All Rights Reserved.
.\" Copyright (c) 2020 Red Hat.  All Rights Reserved.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMDALUSTRE 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdalustre\f1 \- lustre filesystem statistics performance metrics domain agent (PMDA)
.SH DESCRIPTION
\f3pmdalustre\f1 is a Performance Metrics Domain Agent (\s-1PMDA\s0) which
reads and exports metric values from the statistics interfaces of a Lustre filesystem.
.PP
.B pmdalustre
searches for statistics interfaces from several locations,
which vary depending on the version of
.B Lustre
installed locally.
.PP
Recent
.B Lustre
versions (v2.12 and later) export statistics from the kernel
.B debugfs
pseudo filesystem,
in the directories
.B /sys/kernel/debug/lustre/llite
and
.BR /sys/kernel/debug/lnet .
If these interfaces are not found during startup,
.B pmdalustre
will automatically check if the statistics interfaces are available from the
.B procfs
pseudo filesystem below the
.B /proc/fs/lustre/llite
and
.B /proc/sys/lnet
directories.
These are the default locations of the statistics for
.B Lustre
versions less than
.BR v2.12 .
.PP
If neither of the above filesystem interfaces are detected,
or if the user wants to override the default locations,
.B pmdalustre
also supports an optional configuration file named
.BR $PCP_PMDAS_DIR/lustre/lustre.conf .
Note that
.B $PCP_PMDAS_DIR
is set to
.B /var/lib/pcp/pmdas
on most Linux based  systems.
The configuration file supports
.BR perl (1)
variable assignment syntax.
An example configuration file suitable for
.B Lustre
v2.12 and later is:
.in 2i
.nf

.B $LLITE_PATH="/sys/kernel/debug/lustre/llite/";
.B $LNET_PATH="/sys/kernel/debug/lustre/lnet/";

.fi
.in
See comments in the shipped
.B lustre.conf
file for further details.
By default, this file is installed with everything commented
(and so it has no effect unless edited) because the built-in heuristics used by
.B pmdalustre
should suffice.
.PP
Finally, overriding all of the above, the
.B LUSTRE_LLITE_PATH
and
.B LUSTRE_LNET_PATH
environment variables may be set (and exported) to specify the directory locations
of the statistics interfaces to be used.
This mechanism using environment variables is intended to be used
for development and testing purposes only.
.PP
The
.B pmdalustre
process runs as the root user because
.B debugfs
directories are not normally readable by unprivileged users.
.SH INSTALLATION
Install
.B pmdalustre
by using the
.B Install
script as the root user:
.PP
.ft CR
.nf
.in +0.5i
# cd $PCP_PMDAS_DIR/lustre
# ./Install
.in
.fi
.ft 1
.PP
If you want to undo the installation, do the following as root:
.PP
.ft CR
.nf
.in +0.5i
# cd $PCP_PMDAS_DIR/lustre
# ./Remove
.in
.fi
.ft 1
.PP
\fBpmdalustre\fR is launched by \fIpmcd\fR(1) and should never be executed
directly. The Install and Remove scripts notify \fIpmcd\fR(1) when the
agent is installed or removed.
.SH FILES
.IP "\fB$PCP_PMDAS_DIR/lustre/Install\fR" 4
installation script for the \fBpmdalustre\fR agent
.IP "\fB$PCP_PMDAS_DIR/lustre/Remove\fR" 4
undo installation script for the \fBpmdalustre\fR agent
.IP "\fB$PCP_LOG_DIR/pmcd/lustre.log\fR" 4
default log file for error messages from \fBpmdalustre\fR
.IP "\fB$PCP_PMDAS_DIR/lustre/lustre.conf\fR" 4
Configuration file for over-riding default statistics directories used by \fBpmdalustre\fP.
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fR are used to parameterize
the file and directory names used by \fBPCP\fR. On each installation, the
file \fB/etc/pcp.conf\fR contains the local values for these variables.
The \fB$PCP_CONF\fR variable may be used to specify an alternative
configuration file, as described in \fIpcp.conf\fR(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR perl (1),
.BR pmcd (1)
and
.BR llstat (1).

.\" control lines for scripts/man-spell
.\" +ok+ LUSTRE_LLITE_PATH LUSTRE_LNET_PATH LLITE_PATH pmdalustre LNET_PATH
.\" +ok+ debugfs llstat procfs perl
.\" +ok+ sys llite {both from /sys/kernel/debug/lustre/llite}
.\" +ok+ lnet {from /sys/kernel/debug/lnet}
.\" +ok+ fs {from /proc/fs/lustre/llite}


CHANGELOG
pcp-6.2.2 (29 May 2024)
    - Work-in-progress, not yet released.

pcp-6.2.1 (10 April 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdauwsgi: (new) instrumentation from uWSGI servers
	pmdalinux: new hugepage metrics from sysfs (fixed size hugepages)
	pmdalinux: new filesys.uuid and filesys.type metrics
	pmdalinux: new TcpExt metrics from /proc/net/netstat
	pmdalinux: new softnet metrics from newer kernel versions
	pmdalinux: new /proc/vmstat khugepaged metrics
	pmdabpf: updates to add eBPF networking metrics for pcp-atop
	pmdaopenmetrics: fix script error diagnostic cascading exception
    - Client tools and utilities:
	pcp2openmetrics: (new) push PCP metrics in OpenMetrics format
	pmcheck: (new) interrogate and control PCP components
	pmcheck: Redis server detection and agent recommendation
	pmcheck: PostgreSQL server detection and agent recommendation
	pmcheck: uwSGI server detection and agent recommendation
	pcp-atopsar: fix TZ mishandling relating to day-crossover
	pcp-dstat: fix an I/O config typo, sum reads and writes
	pcp-htop: sync latest htop-3.3.0 fixes
	pmrep.conf: use bpf instead of bcc pmda for proc net metrics
	pmrep.conf: proc-essential cosmetic improvements
	pmrep.conf: remove vmeff field from sar -B config
	pmrepconf: use pmrep-internal representations for each metric
    - Server-side utilities and log management scripts:
	pmlogger: bizarre timezone fix (github #1936)
	pmlogger: improve handling of a change in metadata semantics
	pmlogdump: add support for multi-archive contexts
	pmproxy: disable Redis protocol proxying by default
	runaspcp: remove hard-coded "pcp" for user and group
	init scripts: drop conditonal use of setpriv(1) or runuser(1)
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix logmeta.c SEGV with V3 archives
	libpcp: fix memleak on repeated __pmFixPMNSHashTab calls
	libpcp: add pmgetopt pmflush() call for warnings
	libpcp: fixes for V3 archives and multi-archive contexts
	libpcp: export __pmLogChangeArchive() interface to tools
	libpcp: fix for PMNS with multi-archive contexts
	libpcp: small derived metrics bug fix (github #1921)
    - Build, infrastructure and packaging updates:
	Linux MX distro support (Debian without systemd)
	riscv: configure pmdabpf_arch value correctly
	tar packages: fix perl modules install in postinstall script
	eBPF updates: add submodule for blazesym, sync to latest
	debian build: add postrm for pcp-zeroconf
	build: removed cppcheck from 'make check', too flakey
	container: Fedora 40 container testing added
	constainer: fixed Fedora base image location (quay.io)
	build: add support for riscv64 to pcp rpm spec files
    - Security Enhanced Linux:
	selinux policy: small tweak for pmie signal handling
    - Documentation:
	man pages: document pmlogger_check and pmlogger_daily SaveLogs
	man pages: man-spell and corrections
	docs: updated donations page with new fiscal sponsor details

pcp-6.2.0 (12 February 2024)
    - Client tools and utilities:
	pmlogger: default to creating version 3 PCP archives
	pmie, pmlogger, pmcd, pmproxy: init script hardening to
	    run less as root:root, more as pcp:pcp and more safely
	pcp-reboot-init: new systemd service for $PCP_RC_DIR setup
	pcp2arrow: new, export metrics to parquet format for pandas
	pcp-atop: deterministic process sort when handling NULLs
	pcp-atop: update to v2.10.0 upstream code
	pcp-htop: update to latest upstream fixes
	pcp-geolocate: function with older versions of python
	pcp-meminfo: resolve possible 'broken pipe' diagnostics
	pcp-ss: fix handling of listening state sockets to match ss
	pcp-ss: fix the UDP UNCONN vs LISTEN state handling
    - PMDA additions, enhancements and bug fixes:
	pmdaproc: improvements for process accounting metrics
	pmdalinux: add pgpromote* and pgdemote* memory metrics
	pmdalinux: add metric sysfs.module.zswap.max_pool_percent
	pmdalinux: add new SNMP metric network.ip.outtransmits
	pmdasummary: send state change 0 (no change) back to pmcd
	pmdapostgresql: support the psycopg3 python module as well
	pmdaopenmetrics: kepler support - live metrics and logging
    - Server-side utilities and log management scripts:
	pmlogrewrite, pmlogredact: fix for missing data volumes
	pmcd: cosmetic change to avoid pipe2 name clash
	runaspcp: new helper utility for unprivileged rc scripts
	pmnsmerge: be more defensive in qsort comparison routine
	pmlogsize: fix qsort comparison function handling of equality
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: pmOpenLog refactor for handling previous log files
	libpcp: new __pmCleanMapDir routine to aid log culling
	libpcp & pmcd: READY-NOT-READY hardening
	libpcp: avoid possible thread issues (coverity)
	libpcp: refactor pmDupContext for thread safety
	libpcp_import: add guard in pmiPutMark to prevent SIGSEGV
    - Build, infrastructure and packaging updates:
	build: changes to re-enable builds for FreeBSD 12, 13, 14
	build: add diffutils rpm dep as pmdaproc script uses diff
	build: add main.yml for building and publishing container
	build: resolve compiler warnings in various tools/libraries
	packaging: use tmpfiles.d method to create PCP_RUN_DIR
	build: packaging changes for openSUSE 15.5 and 15.6
	debian: do not ship empty directory /usr/lib/pkgconfig
	build: el10 systemd-presets for pmcd/pmie/pmlogger enabling
	build: drop rpm dep on bpftool as local version used only
	build: change perl YAML::XS::LibYAML module to YAML::XS
    - Security Enhanced Linux:
	selinux: policy improvements for CentOS/RHEL 7 and 8
    - Documentation:
	docs: Fedora 39 found some tbl(1)/troff(1) issues, fix
	docs: add an explicit code-of-conduct for PCP
	docs: pmns(5) note that metric names must only contain ASCII
	docs: update the BPF README to describe bpftool reality

pcp-6.1.1 (17 November 2023)
    - Client tools and utilities:
	pmie: add -o/--format to control output format with archives

src/pmdas/lustre/pmdalustre.pl
#
# Copyright (c) 2015 Martins Innus.  All Rights Reserved.
# Copyright (c) 2020 Red Hat.  All Rights Reserved.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#

use strict;
use warnings;
use PCP::PMDA;

our $noval = PM_ERR_APPVERSION;

use vars qw( $pmda );

our  $lustre_llite_indom = 0;

# Directories for llite and lnet stats - defaults for Lustre v2.12 and later
our $LLITE_PATH = "/sys/kernel/debug/lustre/llite/";
our $LNET_PATH = "/sys/kernel/debug/lnet/";

# Fall-back defaults for pre-Lustre v2.12 using the /proc interfaces
if (! -d $LLITE_PATH || ! -d $LNET_PATH) {
    $LLITE_PATH = "/proc/fs/lustre/llite/";
    $LNET_PATH = "/proc/sys/lnet/";
}

# Optional config files for overriding the stats dirs, mostly for testing purposes
for my $file (pmda_config('PCP_PMDAS_DIR') . '/lustre/lustre.conf', 'luster.conf') {
        eval `cat $file` unless ! -f $file;
}

# Optional env variables for overriding the stats dirs to use, also mostly for testing
if ( defined $ENV{"LUSTRE_LLITE_PATH"} ) {
	$LLITE_PATH = $ENV{"LUSTRE_LLITE_PATH"}
}
if ( defined $ENV{"LUSTRE_LNET_PATH"} ) {
	$LNET_PATH = $ENV{"LUSTRE_LNET_PATH"}
}

if ( -d $LLITE_PATH && -d $LNET_PATH ) {
    printf STDERR "Using Lustre stats in '%s' and %s\n", $LLITE_PATH, $LNET_PATH;
}
else {
    print STDERR "FATAL Error: no Lustre stats interfaces detected, see pmdalustre(1).\n";
    exit;
}

# List of metrics we care about

# llite
# Bytes have: min, max, count, total
our @llite_byte_stats = ('read_bytes', 'write_bytes', 'osc_read', 'osc_write');

# Pages have: min, max, count, total
our @llite_page_stats = ('brw_read', 'brw_write');

# Reg have: count
our @llite_reg_stats = ('dirty_pages_hits', 'dirty_pages_misses', 'ioctl', 'open', 'close',
	   'mmap', 'seek', 'fsync', 'readdir', 'setattr',
	   'truncate', 'flock', 'getattr', 'create', 'link', 'unlink',
	   'symlink', 'mkdir', 'rmdir', 'mknod', 'rename',
	   'statfs', 'alloc_inode', 'setxattr', 'getxattr', 'getxattr_hits', 'listxattr',
	   'removexattr', 'inode_permission');



# The stats hash is keyed on base name of the entires in /proc/fs/lustre/llite/ e.g.
# 'lustre' from lustre-ffff880378305c00.  The value is a hash ref(a mount and can have multiple per host), 
# and that hash contains all of the stats data keyed on pmid_name.
our %h_llite = ();

# Only one set of lnet metrics per host
# Still use a ref so its clear theres no indom
our $h_lnet = {};

#
# Find all the lustre file systems on the host
# Devices are of the form "foo-<16_hex_chars>"
#
# And parse the stats file underneath

sub lustre_get_llite_stats{
	opendir(LLITEDIR, $LLITE_PATH) || die "Can't open any stats directories, is Lustre installed?\n";

	while(my $ldev = readdir(LLITEDIR) ){
		if( $ldev =~ /^(\w+)-([a-f0-9]{16})$/ ){

			# Get the device information

			my $mtroot = $1;
			my $hexsuper = $2;
			$h_llite{$mtroot} = {}; # {} is an anonymous hash ref
			$h_llite{$mtroot}->{'lustre.llite.volume'} = $mtroot;
			$h_llite{$mtroot}->{'lustre.llite.superblock'} = $hexsuper;

			# Parse the actual stats file
			# From: lustre-2.5.3/lustre/llite/lproc_llite.c
			# Stats do not exists unless the counter has been incremented. Need to initialize to 0
			
			for my $stat_name (@llite_byte_stats) {
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.count"} = 0;
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.min"} = 0;
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.max"} = 0;
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.total"} = 0;
			}
			for my $stat_name (@llite_page_stats) {
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.count"} = 0;
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.min"} = 0;
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.max"} = 0;
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.total"} = 0;
			}
			for my $stat_name (@llite_reg_stats) {
				$h_llite{$mtroot}->{"lustre.llite.$stat_name.count"} = 0;
			}
			
			my $statspath = $LLITE_PATH . $ldev . '/stats';

			if( ! open STATS, '<', $statspath ) {
                		$pmda->err("pmdalustre failed to open $statspath: $!");
                		die "Can't open $statspath: $!\n";
			}

			while (<STATS>) {
                		my $line = $_;
				# Byte types first
				my $byte_types = join '|', @llite_byte_stats;
				my $pages_types = join '|', @llite_page_stats;
				my $reg_types = join '|', @llite_reg_stats;
				if( $line =~ /^($byte_types)\s+(\d+) samples \[bytes\] (\d+) (\d+) (\d+)$/){
					my $name = $1;
					my $count = $2;
					my $min = $3;
					my $max = $4;
					my $total = $5;
					$h_llite{$mtroot}->{"lustre.llite.$name.count"} = $count;
					$h_llite{$mtroot}->{"lustre.llite.$name.min"} = $min;
					$h_llite{$mtroot}->{"lustre.llite.$name.max"} = $max;
					$h_llite{$mtroot}->{"lustre.llite.$name.total"} = $total;
				# Pages types
				# I have not seen these, so can't test
				} elsif ( $line =~ /^($pages_types)\s+(\d+) samples \[pages\] (\d+) (\d+) (\d+)$/){
					my $name = $1;
                                        my $count = $2;
                                        my $min = $3;
                                        my $max = $4;
                                        my $total = $5;
                                        $h_llite{$mtroot}->{"lustre.llite.$name.count"} = $count;
                                        $h_llite{$mtroot}->{"lustre.llite.$name.min"} = $min;
                                        $h_llite{$mtroot}->{"lustre.llite.$name.max"} = $max;
                                        $h_llite{$mtroot}->{"lustre.llite.$name.total"} = $total;
				# Regs types
				} elsif ( $line =~ /^($reg_types)\s+(\d+) samples \[regs\]$/ ){
					my $name = $1;
                                        my $count = $2;
					$h_llite{$mtroot}->{"lustre.llite.$name.count"} = $count;
				}
			}
			close( STATS );
		}
	}
	closedir( LLITEDIR );
}

# lnet stats. One set per host

sub lustre_get_lnet_stats{

	my $statspath = $LNET_PATH . 'stats';

	if( ! open STATS, '<', $statspath ){
		$pmda->err ("pmdalustre failed to open $statspath: $!");
		die "Can't open $statspath: $!\n";
	}

	while (<STATS>) {
		my $line = $_;
		# From: lustre-2.5.3/lnet/lnet/router_proc.c
		# which claims to be lnet_proc.c
		if( $line =~ /^(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)\s(\d+)$/){
			my $msgs_alloc = $1;
			my $msgs_max = $2;
			my $errors = $3;
			my $send_count = $4;
			my $recv_count = $5;
			my $route_count = $6;
			my $drop_count = $7;
			my $send_length = $8;
			my $recv_length = $9;
			my $route_length = $10;
			my $drop_length = $11;
			$h_lnet->{'lustre.lnet.msgs_alloc'} = $msgs_alloc;
			$h_lnet->{'lustre.lnet.msgs_max'} = $msgs_max;
			$h_lnet->{'lustre.lnet.errors'} = $errors;
			$h_lnet->{'lustre.lnet.send_count'} = $send_count;
			$h_lnet->{'lustre.lnet.recv_count'} = $recv_count;
			$h_lnet->{'lustre.lnet.route_count'} = $route_count;
			$h_lnet->{'lustre.lnet.drop_count'} = $drop_count;
			$h_lnet->{'lustre.lnet.send_length'} = $send_length;
			$h_lnet->{'lustre.lnet.recv_length'} = $recv_length;
			$h_lnet->{'lustre.lnet.route_length'} = $route_length;
			$h_lnet->{'lustre.lnet.drop_length'} = $drop_length;
		}
	}
	close( STATS );
}

#
# fetch is called once by pcp for each refresh and then the fetch callback is
# called to query each statistic individually
#
sub lustre_fetch {
	lustre_get_llite_stats();
	lustre_get_lnet_stats();

	our $pmda->replace_indom($lustre_llite_indom, \%h_llite);
}

sub lustre_fetch_callback {
	my ($cluster, $item, $inst) = @_; 

	if( $cluster == 0 ){
		# llite
		my $lookup = pmda_inst_lookup($lustre_llite_indom, $inst);
		return (PM_ERR_INST, 0) unless defined($lookup);

		my $pmid_name = pmda_pmid_name($cluster, $item)
			or die "Unknown metric name: cluster $cluster item $item\n";

		return ($lookup->{$pmid_name}, 1);
	}
	elsif( $cluster == 1 ){
		# lnet
		my $pmid_name = pmda_pmid_name($cluster, $item)
			or die "Unknown metric name: cluster $cluster item $item\n";

		return ($h_lnet->{$pmid_name}, 1);
	}
	else{
		return (PM_ERR_PMID, 0);
	}
}

# the PCP::PMDA->new line is parsed by the check_domain rule of the PMDA build
# process, so there are special requirements:  no comments, the domain has to
# be a bare number.
#
our $pmda = PCP::PMDA->new('lustre', 134);

# Metrics

# llite stats - cluster 0
# lnet stats - cluster 1

# llite

$pmda->add_metric(pmda_pmid(0,1), PM_TYPE_STRING, $lustre_llite_indom,
		PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		'lustre.llite.volume',
		'Volume Name',
		'');
$pmda->add_metric(pmda_pmid(0,2), PM_TYPE_STRING, $lustre_llite_indom,
		PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		'lustre.llite.superblock',
		'Superblock Identifier',
		'');

my $item = 3;
for my $stat_name (@llite_byte_stats) {

	$pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
			  PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
			  "lustre.llite.$stat_name.count",
			  'number of calls',
			  '');

	$pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
			  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
			  "lustre.llite.$stat_name.min",
			  'minimum byte value seen in a call',
			  '');

	$pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
			  PM_SEM_INSTANT, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
			  "lustre.llite.$stat_name.max",
			  'maximum byte value seen in a call',
			  '');

	$pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
			  PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
			  "lustre.llite.$stat_name.total",
			  'total byte count',
			  '');
}

for my $stat_name (@llite_page_stats) {

        $pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
                          PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                          "lustre.llite.$stat_name.count",
                          'number of calls',
                          '');

        $pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
                          PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
                          "lustre.llite.$stat_name.min",
                          'minimum page value seen in a call',
                          '');

        $pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
                          PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
                          "lustre.llite.$stat_name.max",
                          'maximum page value seen in a call',
                          '');

        $pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
                          PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                          "lustre.llite.$stat_name.total",
                          'total page count',
                          '');
}

for my $stat_name (@llite_reg_stats) {

        $pmda->add_metric(pmda_pmid(0, $item++), PM_TYPE_U64, $lustre_llite_indom,
                          PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                          "lustre.llite.$stat_name.count",
                          'number of calls',
                          '');

}

# lnet
$item = 1;

# Can't loop through, units are all different
$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
		PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
		"lustre.lnet.msgs_alloc",
		'messages currently allocated',
		'');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
                PM_SEM_INSTANT, pmda_units(0,0,0,0,0,0),
                "lustre.lnet.msgs_max",
                'messages maximum',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                "lustre.lnet.errors",
                'errors',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                "lustre.lnet.send_count",
                'transmit count',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                "lustre.lnet.recv_count",
                'receive count',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                "lustre.lnet.route_count",
                'route count',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U32, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(0,0,1,0,0,PM_COUNT_ONE),
                "lustre.lnet.drop_count",
                'drop count',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U64, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
                "lustre.lnet.send_length",
                'transmit bytes',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U64, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
                "lustre.lnet.recv_length",
                'receive bytes',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U64, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
                "lustre.lnet.route_length",
                'route bytes',
                '');

$pmda->add_metric(pmda_pmid(1, $item++), PM_TYPE_U64, PM_INDOM_NULL,
		PM_SEM_COUNTER, pmda_units(1,0,0,PM_SPACE_BYTE,0,0),
                "lustre.lnet.drop_length",
                'drop bytes',
                '');


&lustre_get_llite_stats;
&lustre_get_lnet_stats;

$lustre_llite_indom = $pmda->add_indom($lustre_llite_indom, {}, '', '');

$pmda->set_fetch(\&lustre_fetch);
$pmda->set_fetch_callback(\&lustre_fetch_callback);

$pmda->run;
